echopop/acoustics.py

```diff
@@ -0,0 +1,452 @@
+            configuration_dict["TS_length_regression_parameters"]["pacific_hake"],
+            settings_dict,
+        )
+        # ---- Calculate adult proportions
+        # -------- Initialize dataframe
+        adult_proportions = age1_proportions.copy()
+        # -------- Invert proportions
+        adult_proportions.iloc[:, np.arange(1, 4)] = (
+            1.0 - adult_proportions.iloc[:, np.arange(1, 4)]
+        )
+
+    else:
+        # ---- Drop the unused column
+        nasc_interval_df.drop("NASC_no_age1", axis=1, inplace=True)
+        # ---- Rename the used column
+        nasc_interval_df.rename(columns={"NASC_all_ages": "nasc"}, inplace=True)
+        # ---- Assign filled adult proportions dataframe
+        adult_proportions = pd.DataFrame(
+            {
+                f"{stratum_col}": np.unique(nasc_interval_df[stratum_col]),
+                "number_proportion": np.ones(len(np.unique(nasc_interval_df[stratum_col]))),
+                "weight_proportion": np.ones(len(np.unique(nasc_interval_df[stratum_col]))),
+                "nasc_proportion": np.ones(len(np.unique(nasc_interval_df[stratum_col]))),
+            }
+        )
+
+    # Merge hake fraction data into `nasc_interval_df`
+    # ---- Initial merge
+    nasc_interval_df = nasc_interval_df.merge(
+        input_dict["spatial"]["strata_df"], on=[stratum_col, "haul_num"], how="outer"
+    )
+    # ---- Replace `fraction_hake` where NaN occurs
+    nasc_interval_df["fraction_hake"] = nasc_interval_df["fraction_hake"].fillna(0.0)
+    # ---- Drop NaN
+    nasc_interval_df.dropna(subset=["transect_num"], inplace=True)
+
+    # Calculate the along-transect number density (animals per nmi^2)
+    # ---- Merge NASC measurements with mean sigma_bs for each stratum
+    nasc_biology = nasc_interval_df.merge(sigma_bs_strata, on=[stratum_col])
+    # ---- Calculate the number densities
+    nasc_biology["number_density"] = (
+        nasc_biology["fraction_hake"]
+        * nasc_biology["nasc"]
+        / (4.0 * np.pi * nasc_biology["sigma_bs_mean"])
+    )
+    # ---- Round to the nearest whole number (we don't want fractions of a fish)
+    nasc_biology["number_density"] = np.round(nasc_biology["number_density"])
+
+    # Calculate the along-transect abundance (# animals)
+    nasc_biology["abundance"] = nasc_biology["interval_area"] * nasc_biology["number_density"]
+
+    # Calculate the along-transect abundance (#, N) and biomass for each sex (kg)
+    # ---- Extract the sex-specific number proportions for each stratum
+    sex_stratum_proportions = analysis_dict["biology"]["proportions"]["number"][
+        "sex_proportions_df"
+    ].copy()
+    # ---- Filter out sexes besides 'male' and 'female'
+    sex_stratum_proportions = sex_stratum_proportions[
+        sex_stratum_proportions.sex.isin(["male", "female"])
+    ]
+    # ---- Merge with the NASC measurements
+    nasc_biology_sex = nasc_biology.merge(sex_stratum_proportions, on=[stratum_col, "species_id"])
```

Comment:
I may be missing something but I don't think it makes sense to separate out different species here. In the above you already have `hake_fraction`, which is hard coded to a single species. I have the same suggestion as in the function that calculates mean sigma_bs that it is simpler to just handle one species.

Reply:
I see your point since both this function the (previously known called) `summarize_sigma_bs(...)` can just as easily applied via `.groupby[["other contrasts", "species_id"]]` rather than have those ingrained within the individual functions. I suppose that results in one less thing to check for in terms of bugs, errors, etc?